### ISSUE TITLE
Flux.cache(Duration) is not expiring after ttl passed

### ISSUE THREAD

**Issue by mkadan** [reactor/reactor-core#750]
### Expected behavior
Flux.cache(Duration) should replay only during ttl.

### Actual behavior
It replays indefinitely.

### Steps to reproduce
```
AtomicInteger i = new AtomicInteger(0);
Flux<Integer> connect = Flux.defer(() -> {
    System.out.println("refresh");
    return Flux.just(i.getAndIncrement());
})
        .cache(Duration.ofMillis(1000));

connect.subscribe(System.out::println);
connect.subscribe(System.out::println);

Thread.sleep(2000);
connect.subscribe(System.out::println);
connect.subscribe(System.out::println);

Thread.sleep(2000);
connect.subscribe(System.out::println);
connect.subscribe(System.out::println);
```

expected output:
```
refresh
0
0
refresh
1
1
refresh
2
2
```

actual output:
```
refresh
0
0
0
0
0
0
```

### Reactor Core version
3.0.7.RELEASE and 3.1.0.M3

### JVM version (e.g. `java -version`)
1.8.0_65

### OS version (e.g. `uname -a`)
Windows 10

**Comment by smaldini** [comment]
Right now it does only expire elements but not completion altogether, I do think its a valuable enhancement tho so I've experimented a fix, look for the PR and 3.1.0.B-S once accepted.

**Comment by mkadan** [comment]
Thx for looking into it!

If I got it right, currently cache() behaves as I expect it to only before onComplete() is called on the Publisher, after that it will never expire?

I assembled a small snippet to check that suggestion, but seems I'm wrong.

```
AtomicInteger i = new AtomicInteger(0);
Flux<Integer> connect = Flux.defer(() -> (Publisher<Integer>) s -> s.onSubscribe(new Subscription() {

    @Override
    public void request(long n) {
        LOG.info("refresh");
        s.onNext(i.getAndIncrement());
    }

    @Override
    public void cancel() {
    }
}))
        .cache(Duration.ofMillis(1000))
        .log();

connect.subscribe(LOG::info);
connect.subscribe(LOG::info);

Thread.sleep(2000);

connect.subscribe(LOG::info);
connect.subscribe(LOG::info);

Thread.sleep(2000);

connect.subscribe(LOG::info);
connect.subscribe(LOG::info);
```

Again, I was expecting to see:
```
refresh
0
0
refresh
1
1
refresh
2
2
```

but actually got:
```
refresh
0
0
```

Looks like elements were completely dropped after the cache expired.

When I enable logging I see:
```
2017-07-22 14:29:57.877  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:29:57.879  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
2017-07-22 14:29:57.882  INFO 19320 --- [           main] c.e.reactor.service.ReactiveService      : request
2017-07-22 14:29:57.882  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onNext(0)
2017-07-22 14:29:57.882  INFO 19320 --- [           main] c.e.reactor.service.ReactiveService      : 0
2017-07-22 14:29:57.882  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:29:57.882  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
2017-07-22 14:29:57.882  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onNext(0)
2017-07-22 14:29:57.882  INFO 19320 --- [           main] c.e.reactor.service.ReactiveService      : 0
2017-07-22 14:29:59.884  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:29:59.884  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
2017-07-22 14:29:59.885  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:29:59.886  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
2017-07-22 14:30:01.886  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:30:01.887  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
2017-07-22 14:30:01.887  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | onSubscribe([Fuseable] FluxReplay.ReplayInner)
2017-07-22 14:30:01.887  INFO 19320 --- [           main] reactor.Flux.AutoConnectFuseable.1       : | request(unbounded)
```

Could you please provide some info in that matter, just for better understanding of the concept?
